bob/src/hw_metrics_collector.rs

```diff
@@ -473,50 +474,63 @@ impl DescrCounter {
         }
     }
 
-    fn descr_amount(&mut self) -> u64 {
+    async fn descr_amount(&mut self) -> u64 {
         if self.cached_times == 0 {
             self.cached_times = CACHED_TIMES;
-            self.value = self.count_descriptors();
+            self.value = self.count_descriptors().await;
         } else {
             self.cached_times -= 1;
         }
         self.value
     }
 
-    fn count_descriptors_by_lsof(&mut self) -> Option<u64> {
+    async fn count_descriptors_by_lsof(&mut self) -> Option<u64> {
         if !self.lsof_enabled {
             return None;
         }
 
-        let pid_arg = process::id().to_string();
+        let pid_arg = std::process::id().to_string();
         let cmd_lsof = Command::new("lsof")
             .args(["-a", "-p", &pid_arg, "-d", "^mem", "-d", "^cwd", "-d", "^rtd", "-d", "^txt", "-d", "^DEL"])
             .stdout(std::process::Stdio::piped())
             .spawn();
         match cmd_lsof {
-            Ok(cmd_lsof) => {
-                match cmd_lsof.stdout {
+            Ok(mut cmd_lsof) => {
+                match cmd_lsof.stdout.take() {
                     Some(stdout) => {
-                        match parse_command_output(Command::new("wc").arg("-l").stdin(stdout)) {
-                            Ok(output) => {
-                                match output.trim().parse::<u64>() {
-                                    Ok(count) => {
-                                        return Some(count - 5); // exclude stdin, stdout, stderr, lsof pipe and wc pipe
-                                    }
+                        match TryInto::<std::process::Stdio>::try_into(stdout) {
+                            Ok(stdio) => {
+                                match parse_command_output(Command::new("wc").arg("-l").stdin(stdio)).await {
+                                    Ok(output) => {
+                                        match output.trim().parse::<u64>() {
+                                            Ok(count) => {
+                                                if let Err(e) = cmd_lsof.wait().await {
+                                                    debug!("lsof exited with error {}", e);
+                                                }
+                                                return Some(count - 5); // exclude stdin, stdout, stderr, lsof pipe and wc pipe
+                                            }
+                                            Err(e) => {
+                                                debug!("failed to parse lsof result: {}", e);
+                                            }
+                                        }
+                                    },
                                     Err(e) => {
-                                        debug!("failed to parse lsof result: {}", e);
+                                        debug!("can't use lsof, wc error (fs /proc will be used): {}", e);
```

Comment:
Why are the error messages swapped?

Reply:
There is another level of match added

Reply:
Good. Git diff a little bit misleading here. In the whole file I see, that everything is correct.

The level of nesting became very large, but I did not find an easy way to refactor it, since we process each error independently